Add missing headers

etcwilde opened the pull request:
Clang CachedDiagnostics uses std::unique_ptr, std::variant, and std::optional, coming from the `memory`, `variant`, and `optional` headers respectively. It looks like things shifted a bit so this wasn't building locally on Linux due to missing definitions.

llvm subtarget parser SubtargetFeature.h uses `llvm::popcount`, which comes from `llvm/ADT/bit.h`, which was causing build failures locally due to a missing `llvm::popcount` declaration.

Add missing `optional` include to include/llvm/Support/Format.h

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

edymtt reviewed (approved):
Thanks! That allowed me to clear errors related to `SwiftRemoteMirror.cpp` when building Swift for macOS against Xcode 14.2

etcwilde commented:
https://github.com/apple/swift/pull/67170

@swift-ci please test

etcwilde commented:
CMake version we're using is too old for LLVM:

```
CMake Error at CMakeLists.txt:3 (cmake_minimum_required):
  CMake 3.20.0 or higher is required.  You are running version 3.19.6
```

etcwilde commented:
Alright, this gets us through the LLVM build and into Swift build failures due to leaky builds;

```
In file included from /Users/ec2-user/jenkins/workspace/apple-llvm-project-pr-macos/branch-next/swift/stdlib/public/SwiftRemoteMirror/SwiftRemoteMirror.cpp:26:
In file included from /Users/ec2-user/jenkins/workspace/apple-llvm-project-pr-macos/branch-next/swift/include/swift/RemoteInspection/ReflectionContext.h:24:
In file included from /Users/ec2-user/jenkins/workspace/apple-llvm-project-pr-macos/branch-next/llvm-project/llvm/include/llvm/Object/COFF.h:26:
In file included from /Users/ec2-user/jenkins/workspace/apple-llvm-project-pr-macos/branch-next/llvm-project/llvm/include/llvm/TargetParser/SubtargetFeature.h:20:
/Users/ec2-user/jenkins/workspace/apple-llvm-project-pr-macos/branch-next/llvm-project/llvm/include/llvm/ADT/bit.h:115:51: error: redefinition of 'TrailingZerosCounter'
  115 | template <typename T, std::size_t SizeOfT> struct TrailingZerosCounter {
      |                                                   ^
/Users/ec2-user/jenkins/workspace/apple-llvm-project-pr-macos/branch-next/swift/stdlib/include/llvm/Support/MathExtras.h:91:51: note: previous definition is here
   91 | template <typename T, std::size_t SizeOfT> struct TrailingZerosCounter {
      |                                                   ^
```

(I think this is as good as we're going to get for a bit.)
I've opened https://reviews.llvm.org/D154664 on upstream LLVM so we'll get that landed there.

etcwilde committed:
[NFC] Add missing Clang Cache headers

Clang cache uses std::variant, std::unique_ptr, and std::optional. The
missing headers are resulting in build failures.

etcwilde committed:
[NFC] Add missing bit.h to SubtargetFeature.h

SubtargetFeature.h uses llvm/ADT/bit.h.

Reviewed By: compnerd, mib

Differential Revision: https://reviews.llvm.org/D154664

etcwilde commented:
@swift-ci please smoke test
